thriftbp/server_middlewares.go

```diff
@@ -299,16 +299,22 @@ func ReportPayloadSizeMetrics(rate float64) thrift.ProcessorMiddleware {
 							oproto.Flush(ctx)
 
 							proto := "header-" + tHeaderProtocol2String(protoID)
+							labels := prometheus.Labels{
+								methodLabel: name,
+								protoLabel:  proto,
+							}
 							metricsbp.M.HistogramWithRate(metricsbp.RateArgs{
 								Name:             "payload.size." + name + ".request",
 								Rate:             1,
 								AlreadySampledAt: metricsbp.Float64Ptr(rate),
 							}).With("proto", proto).Observe(float64(itrans.Size()))
+							payloadSizeRequestBytes.With(labels).Observe(float64(itrans.Size()))
```

Comment:
This section needs to be restructured a bit.

`if randbp.ShouldSampleWithRate(rate) {` above needs to be moved to wrap around the `metricsbp.M.HistogramWithRate()` calls so that we can use `prometheus.Observe()` on every call, but only sample for statsd calls.

Reply:
that means we need to apply the overhead of counting the bytes to all requests. we don't want to do that (for now). we might revisit that after we removed statsd metrics.

Reply:
Let's leave this out of the change for now. We do not want to introduce pre-sampled histograms into Prometheus metrics. It's just not handled by the ecosystem.

Reply:
this is an important metrics that we don't want to lose when we migrate to prometheus. also I don't think it will break the system in any way? we just have randomly sampled reporting instead of reporting all, which does not really have any real impact on p99/p50 in statistical ways?

Reply:
We need to defer to the expert here, and the clear message is that we do not sample prometheus metrics, full stop.

If you are concerned about performance, I recommend making that decision based on a benchmark or a load test.  If the data says that it's expensive, it should be opt-in (or maybe opt-out), but still recording samples at 100% when enabled.  Either collecting the data is worth the cost or it's not.